File: Golang/740.delete-and-earn.go
/*
 * @lc app=leetcode.cn id=740 lang=go
 *
 * [740] 删除并获得点数
 *
 * 借助一个 sum(x) 数组来辅助处理 - 类似与198.house-robber
 */

package leetcode

// @lc code=start
func deleteAndEarn(nums []int) int {
    maxVal := 0
    for _, val := range nums {
        maxVal = max(maxVal, val)
    }

    sum_nums := make([]int, maxVal+1)
    for _, val := range nums {
        sum_nums[val] += val
    }

    return rob(sum_nums)
}

func rob(nums []int) int {
    first, second := nums[0], max(nums[0], nums[1])
    for idx := 2; idx < len(nums); idx += 1 {
        first, second = second, max(first + nums[idx], second)
    }
    return second
}

func max(a, b int) int {
    if a > b {
        return a
    }
    return b
}
// @lc code=end

File: Cplusplus/740.delete-and-earn.cpp
/*
 * @lc app=leetcode.cn id=740 lang=cpp
 *
 * [740] 删除并获得点数
 *
 * 借助一个 sum(x) 数组来辅助处理 - 类似与198.house-robber
 */

#include <vector>
#include <algorithm>
using namespace std;

// @lc code=start
class Solution
{
public:
    int deleteAndEarn(vector<int> &nums)
    {
        int n = nums.size();
        int max_value = *max_element(nums.begin(), nums.end());
        vector<int> sum_nums = vector<int>(max_value + 1, 0);
        vector<int> tempVec = vector<int>(max_value + 2, 0);

        // calc the sum(x)
        for (int idx = 0; idx < n; ++idx){
            sum_nums[nums[idx]] += nums[idx];
        }

        // regressive
        tempVec[1] = sum_nums[0];
        for (int idx = 2; idx <= max_value + 1; ++idx) {
            tempVec[idx] = max(tempVec[idx - 2] + sum_nums[idx - 1], tempVec[idx - 1]);
        }

        return tempVec[max_value + 1];
    }
};
// @lc code=end
